### DISCUSSION TITLE
AVA 4 removing callback support?

### DISCUSSION THREAD

**Discussion by Devwulf** [avajs/ava#2752] [Q&A] [answered]
While going through the AVA docs found [here](https://github.com/avajs/ava/blob/main/docs/01-writing-tests.md#callback-support), it says that

> AVA 4 removes support for test.cb() and t.end().

Searching in Google about this did not turn up any results, although I didn't search too far. I'm curious, what is the reason for this? What will be the alternatives for callback testing?

**Comment by novemberborn** [comment, marked as answer]
Yes, it's being removed. Modern APIs tend to use promises, not callbacks, and those are best tested using `async`/`await`. Callback tests required a lot of duplication in our type definitions. It didn't seem worth the maintenance cost.

You can use `promisify` from the built-in `util` module to wrap a callback-taking API, or the test implementation itself:

```js
import { promisify } from 'node:util'
import test from 'ava'

const withCallback = fn => async t => {
  await promisify(fn)(t)
  t.pass() // There must be at least one passing assertion for the test to pass
}

test('callbackish', withCallback((t, end) => {
  someCallbackThing(end)
})
```

(I haven't tested this.)

**Reply by Devwulf** [reply]
Thank you for the answer! I was just wondering why, because I myself don't know the current state of APIs and libraries, and the best practices that go with them.

**Comment by joel-chu** [comment]
So how are we going to test a socket app?

```
socket.on('message', function message(msg) {
    // assertion 
})
```

**Reply by sindresorhus** [reply]
https://github.com/sindresorhus/p-event or `new Promise`.
